### ISSUE TITLE
Multi-GPUs training error!! RuntimeError: Expected to mark a variable ready only once. 

### ISSUE THREAD

**Issue by LANYAN1212** [haotian-liu/LLaVA#329]
### Question

Hi! I am using the script of finetune_lora.sh in order to fine-tune LLAVA with lora-llama, and it works well on single GPU. However, when I am using multiple GPUs (8 GPUs), I face this issue and I cannot solve.

RuntimeError: Expected to mark a variable ready only once. This error is caused by one of the following reasons: 1) Use of a module parameter outside the `forward` function. Please make sure model parameters are not shared across multiple concurrent forward-backward passes. or try to use _set_static_graph() as a workaround if this module graph does not change during training loop.2) Reused parameters in multiple reentrant backward passes. For example, if you use multiple `checkpoint` functions to wrap the same part of your model, it would result in the same set of parameters been used by different reentrant backward passes multiple times, and hence marking a variable ready multiple times. DDP does not support such use cases in default. You can try to use _set_static_graph() as a workaround if your module graph does not change over iterations. Parameter at index 445 with name base_model.model.model.layers.31.mlp.down_proj.lora_B.default.weight has been marked as ready twice. This means that multiple autograd engine hooks have fired for this particular parameter during this iteration.

<img width="1485" alt="企业微信截图_16907372202099" src="https://github.com/haotian-liu/LLaVA/assets/56435130/8f746d9f-6167-4cc4-a522-b2dc36461aa7">

**Comment by haotian-liu** [comment]
Hi, can you provide the scripts that you use? Are you using the latest [`finetune_lora.sh`](https://github.com/haotian-liu/LLaVA/blob/main/scripts/finetune_lora.sh)? Thanks.

**Comment by LANYAN1212** [comment]
Yes, this is my script.

torchrun --nproc_per_node=8 --master_port=20001 llava/train/train_mem_LLaVALoRA.py \
    --lora_enable True \
    --model_name_or_path /group/30098/xxx/Projects/LLM/LLMSD/models/vicuna-7b-v1 \
    --version v1 \
    --data_path /group/30042/public_datasets/llava_instruct/llava_instruct_150k.json \
    --image_folder /group/30042/public_datasets/coco/train2017 \
    --vision_tower openai/clip-vit-large-patch14 \
    --pretrain_mm_mlp_adapter ./llava/train/mm_projector.bin \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --bf16 True \
    --output_dir ./LLaVA_finetune_lora_1 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 16 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 50000 \
    --save_total_limit 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --lazy_preprocess True \
    --dataloader_num_workers 4 \
    --report_to wandb

**Comment by LANYAN1212** [comment]
And I am using the latest finetune_lora.sh. I meet this problem.

**Comment by haotian-liu** [comment]
@LANYAN1212 

It seems that you are not using DeepSpeed and using torchrun instead. LoRA is now only tested with DeepSpeed as PyTorch FSDP is known to have issues with PEFT. Can you try with the Deepspeed command in [finetune_lora.sh](https://github.com/haotian-liu/LLaVA/blob/main/scripts/finetune_lora.sh)? Thanks.

**Comment by LANYAN1212** [comment]
Yes! After I use DeepSpeed, the error is fixed. Thank you very much.
By the way, I would like to ask that when I train the model by HF-trainer, since DeepSpeed would automatically convert all data and model into torch.bfloat16 in training loop, I wonder how should I do if I would like to keep data and training parameters on torch.float32 in order to have better training results. Many thanks if you could tell me.

**Comment by haotian-liu** [comment]
I think you can the parameters of enabling bf16 both in command line (`--bf16`) and in deepspeed config. But I have not fully verified this training as it shall double the memory consumption.

**Comment by LANYAN1212** [comment]
Thank you very much for your reply. I would like to have a try.

**Comment by LANYAN1212** [comment]
Maybe I would like to ask the thing that might not so related to LLaVA, when I use torchrun --nproc_per_node=8 to start the code, then I will meet 'Parameter xxx with name lora_A/lora_A.default.weight has been marked as ready twice', I wonder if I use accelerate to train the model, will I meet that problem or not. Thank you very much for that attention.

**Comment by haotian-liu** [comment]
Hi, @LANYAN1212 

Pytorch FSDP is not compatible with PEFT yet. So please stick to DeepSpeed if you want to use LoRA/QLoRA and follow the official sample scripts for the setup. If there is any reason that prevents you from using DeepSpeed, please let me know, thanks.

**Comment by LANYAN1212** [comment]
1. When I am training LLaVA with LLaMA+LoRA, I use DeepSpeed with zero3/zero3_offload.json and I add --bf16 True, then I can train LLaVA successfully, and in the training loop(forward), all model and data will convert into torch.bfloat16. Then, I just set bf16='auto' in both zero3/zero3_offload.json and still keep --bf16 True for DeepSpeed training, then in the training loop, the model will become mixed precision (LLaMA bf16, vision_tower fp16, mm_projector fp32), and the data images become fp32, then the training goes wrong. I wonder which command dominate the data type (--bf16 True for training command or bf16 enable=True in json).
2. I would like to combine LLM+LoRA with stable diffusion, and I would like to finetune unet. Therefore, I would like to conduct the mixed precision training, and I need to put LLM+LoRA/vae on bf16, and unet on fp32, how should I do for both training script(--bf16) and zero3/zero3_offload.json files.
I am very thankful for your attention.

**Comment by haotian-liu** [comment]
1. I would suggest keep deepspeed config and the command line consistent, as the behavior of mismatched config is not guaranteed.
2. I haven't done this myself, but one may try casting the model dtype within the code (e.g., `llm.to(dtype=torch.bfloat16)`), and keep the command line and deepspeed using fp32.

**Comment by haotian-liu** [comment]
I would suggest try figuring these configuration out using a small LLM (e.g. OpenLLaMA-3B), and other components that you can find which is as small as possible. First try to figure the training out in zero2, then to zero3. Zero3 according to my experience can have more compatibility issue to tackle with than zero2. So this process may help you better isolate the issue.

**Comment by LANYAN1212** [comment]
Great! After I switch the zero3_offload to zero2_offload , and I could conduct the training successfully. Thank you very much for your kind suggestion. However, when I conduct inference, I face another problem: 'RuntimeError: GET was unable to find an engine to execute this computation'. I feel very confused. I just save the model in ".bin" type, and I am using torch to load the model. But no matter I use python or deepspeed to run the code, I will always face this issue. Would you have some experiences or advise on it?

**Comment by BlueBlueFF** [comment]
> Thank you very much for your kind suggestion. According to your suggestion, I would like to combine LLM+LoRA with stable diffusion without mixed precision first, which means that I will keep bf16 enabled=true in zero3_offload.json. Thus, all model and data will be put on torch.bfloat16 in forward training loop. However, when I use follow this setting I mentioned above, I always meet 'RuntimeError: still have inflight params', even though the forward training loop can go straight successfully, the error will be met on deepspeed engine optimizer loss backward, and it is hard to solve since there are no suitable solution online. I feel very confused now, since for LLaVA keep bf16 enabled=true in zero3_offload.json could be okay, but I am stuck here. Would you have some experiences or advise on it? Thank you for your attention again.

Same Question.Do you fix it?
